File: leetcode/0213.go
package leetcode

func rob(nums []int) int {
	n := len(nums)
	if n == 1 {
		return nums[0]
	}
	if n == 2 {
		return max(nums[0], nums[1])
	}
	f1, f2, s1, s2 := 0, 0, nums[0], nums[1]
	for i := 1; i < n-1; i++ {
		f1, s1 = s1, max(nums[i]+f1, s1)
		f2, s2 = s2, max(nums[i+1]+f2, s2)
	}
	return max(s1, s2)
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: leetcode/0213.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        int n = nums.size();
        if (n == 1) return nums[0];
        auto f1 = 0, f2 = 0, s1 = nums[0], s2 = nums[1];
        for (int i = 1; i < n - 1; i++) {
            auto t1 = s1, t2 = s2;
            s1 = max(f1 + nums[i], s1);
            s2 = max(f2 + nums[i + 1], s2);
            f1 = t1;
            f2 = t2;
        }
        return max(s1, s2);
    }
};